### PR TITLE
impl(GCS+gRPC): wrap streaming read RPC for storage

### PR DESCRIPTION
**Pull request by coryan**
Introduce a new class that plays the `*Connection` role for the `AsyncReader` class. Unlike the raw streaming read RPCs, this returns the types we want to expose to the applications (e.g. `storage::ObjectMetadata` and `ReadPayload`) as opposed to the raw protos. It will be used in tests mocking the `AsyncReader`, so it must be public.

Part of the work for #12116

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12681)

### CODE REVIEW

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/internal/async/reader_connection_impl.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  future<ReadResponse> Read() override;
+
+ private:
+  future<ReadResponse> OnRead(absl::optional<ProtoPayload> r);
+
+  future<ReadResponse> DoFinish();
+
+  google::cloud::internal::ImmutableOptions options_;
+  std::unique_ptr<StreamingRpc> impl_;
+  std::int64_t offset_ = 0;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/async/reader_connection_impl.h:55]: Done.

File: google/cloud/storage/async_reader_connection.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Asks for more data.
+   *
+   * Retrieving more data can result in three outcomes:
+   * - Additional data (a `ReadPayload`) is available.
+   * - The download is interrupted with an error.
+   * - The download has completed successfully.
+   *
+   * A `StatusOr<>` cannot represent the last outcome, we need an
+   * `absl::variant<>` in this case.  We could have used
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/async_reader_connection.h]: Done.  Though the only people making calls on these objects should be us.  There are too many ways to use them wrong.  "Soon" I will add the class that is hard to use wrong and is intended as a public API.

File: google/cloud/storage/async_reader_connection.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * The `*Connection` object for `AsyncReader`.
+ *
+ * Applications should have little need to use this class directly. They should
+ * use `AsyncReader` instead, which provides an easier to use interface.
+ *
+ * In tests, this class can be used to mock the behavior of `AsyncReader`.
+ */
+class AsyncReaderConnection {
+ public:
+  /// The value returned by `Read()`, see the function for more details.
```
Comment (devbww) [inline google/cloud/storage/async_reader_connection.h:39]: Nit: s/, see/. See/

File: google/cloud/storage/async_reader_connection.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+   * Asks for more data.
+   *
+   * An outcome with a `Status` indicates that no more data is avaiable. Calling
+   * `Read()` after it returns a `Status` results in undefined behavior.
+   *
+   * Applications should not have more than one `Read()` pending at a time.
+   * Calling `Read()` while a previous `Read()` is pending results in undefined
+   * behavior.
+   *
+   * Applications should not destruct a `AsyncReaderConnection` until a call to
```
Comment (devbww) [inline google/cloud/storage/async_reader_connection.h:62]: Nit: s/a Async/an Async/

File: google/cloud/storage/async_reader_connection.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+   *
+   * Callers should continue reading until `Read()` is satisfied with a
+   * `Status`.
+   */
+  virtual void Cancel() = 0;
+
+  /**
+   * Asks for more data.
+   *
+   * An outcome with a `Status` indicates that no more data is avaiable. Calling
```
Comment (devbww) [inline google/cloud/storage/async_reader_connection.h:55]: Nit: s/avaiable/available/

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/async_reader_connection.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/async_reader_connection.h]: Fixed.

File: google/cloud/storage/async_reader_connection.h (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/async_reader_connection.h]: Fixed.

File: google/cloud/storage/async_reader_connection.h (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/async_reader_connection.h]: Fixed.

**Pull request merged**